### PR TITLE
FEAT-#7337: Using dynamic partitionning in broadcast_apply

### LINKED ISSUE #7337: Using dynamic partitioning for broadcast_apply

### PR DESCRIPTION
**Pull request by Retribution98**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #7337? 
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date 

Performance results were obtained by using Dataframe with shape (20000, 5000).
| Operation        | Modin main | This PR |
| ---------------- | ------------- | --------- |
| fillna(df_value) | 32.460861    | 1.79279 |

### CODE REVIEW

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 1]
```diff
@@ -647,6 +709,63 @@ def base_map_partitions(
...
+        """
+        # The condition for the execution of `base_broadcast_apply` is different from
+        # the same condition in the `map_partitions`, since the columnar partitioning approach
+        # cannot be implemented for the `broadcast_apply`. This is due to the fact that different
+        # partitions of the left and right dataframes are possible for the `broadcast_apply`,
+        # as a result of which it is necessary to merge partitions on both axes at once,
+        # which leads to large slowdowns.
+        if (
+            np.prod(left.shape) <= 1.5 * CpuCount.get()
+            or left.shape[axis] < CpuCount.get() // 5
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:749]: Since we build axis partitions with `axis ^ 1`, should we check for `left.shape[axis ^ 1]` here?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 2]
```diff
@@ -647,6 +709,63 @@ def base_map_partitions(
...
+            # block-wise broadcast
+            new_partitions = cls.base_broadcast_apply(
+                axis,
+                apply_func,
+                left,
+                right,
+            )
+        else:
+            # axis-wise broadcast
+            new_partitions = cls.broadcast_axis(
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:760]: Maybe we should rename the current `broadcast_axis_partitions` to `broadcast_right_full` and name `broadcast_axis` as `broadcast_right_axis_paritions`?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 3]
```diff
@@ -492,6 +492,68 @@ def map_func(df, *others):
             ]
         )
 
+    @classmethod
+    @wait_computations_if_benchmark_mode
+    def broadcast_axis(
```
Comment (arunjose696) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:497]: Wouldnt it be possible to  reuse `broadcast_axis_partitions` itself if we just add a additional parameter eg `broadcast_right` which defaults to `false` and sends` right_partitions[i] ` to apply if  this is true.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 1 above)
Comment (Retribution98) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: It has been outdated since the last commits.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 2 above)
Comment (Retribution98) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: I suggest to use below way

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 4]
```diff
@@ -3107,14 +3107,8 @@ def dropna(self, **kwargs):
             lib.no_default,
             None,
         )
-        # FIXME: this is a naive workaround for this problem: https://github.com/modin-project/modin/issues/5394
-        # if there are too many partitions then all non-full-axis implementations start acting very badly.
-        # The here threshold is pretty random though it works fine on simple scenarios
-        processable_amount_of_partitions = (
-            self._modin_frame.num_parts < CpuCount.get() * 32
-        )
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:3115]: Why is this removed?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 5]
```diff
@@ -647,6 +652,56 @@ def base_map_partitions(
...
+
+        Returns
+        -------
+        np.ndarray
+            NumPy array of result partition objects.
+        """
+        # The `broadcast_apply` runtime condition differs from
+        # the same condition in `map_partitions` because the columnar
+        # approach for `broadcast_apply` results in a slowdown.
+        if np.prod(left.shape) <= 1.5 * CpuCount.get():
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:686]: Now we have the same condition, haven't we?

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 6]
```diff
@@ -503,6 +503,7 @@ def broadcast_axis_partitions(
         keep_partitioning=False,
         num_splits=None,
         apply_indices=None,
+        send_all_right=True,
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:506]:
```suggestion
        broadcast_all=True,
```

File: modin/core/dataframe/pandas/partitioning/partition_manager.py [hunk 7]
```diff
@@ -531,6 +532,8 @@ def broadcast_axis_partitions(
             then the number of splits is preserved.
         apply_indices : list of ints, default: None
             Indices of `axis ^ 1` to apply function over.
+        send_all_right : bool, default: True
+            Whether or not to pass all right axis partitions to each of the left axis partitions.
```
Comment (YarShev) [inline modin/core/dataframe/pandas/partitioning/partition_manager.py:536]:
```suggestion
        broadcast_all : bool, default: True
            Whether to broadcast all right axis partitions to each of the left axis partitions or not.
```

**Comment by YarShev** [comment]
Can you extend the table with measurements in the PR description? dropna, groupby?

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 4 above)
Comment (Retribution98) [reply modin/core/storage_formats/pandas/query_compiler.py:3115]: As you can see from this comment, this condition is a naive workaround for a problem that is solved by dynamic partitioning in this PR.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 5 above)
Comment (Retribution98) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:686]: Not really, because in `map_partitions` we use 3 approaches: `base_map_partitions`, `map_partitions_joined_by_column` and `map_axis_partitions`.
In `broadcast_apply` we use only 2 approaches: `base_broadcast_apply` and `broadcast_axis_partitions` because `broadcast_joined_by_axis` does not work as well as for map.

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 6 above)
Comment (Retribution98) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py]: ok, done

**Comment by Retribution98** [comment]
# Actual result from this PR
### fillna
| rows\cols | 32        | 64        | 128       | 256       | 512       | 1024      | 2048      | 3584     |
|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|----------|
| 10**4     | 0.0497    | 0.1502    | 0.1629    | 0.1835    | 0.2736    | 0.4397    | 0.7819    | 1.4975   |
| 10**5     | 0.0454    | 0.1533    | 0.1745    | 0.2250    | 0.3691    | 0.5706    | 1.0826    | 2.0390   |
| 10**6     | 0.1062    | 0.2522    | 0.4855    | 1.0331    | 2.0359    | 5.6754    | 10.8838   | 24.7736  |

### dropna
| rows\cols | 32        | 64        | 128       | 256       | 512       | 1024      | 2048      | 3584     |
|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|----------|
| 10**4     | 0.2251    | 0.1471    | 0.1720    | 0.2370    | 0.3101    | 0.5641    | 1.2769    | 1.9646   |
| 10**5     | 0.2431    | 0.1778    | 0.1855    | 0.2301    | 0.3822    | 0.5796    | 1.1278    | 1.8591   |
| 10**6     | 0.2303    | 0.5102    | 0.5533    | 0.7130    | 1.0950    | 1.9358    | 4.0412    | 6.7829   |
| 10**7     | 0.2371    | 4.8146    | 5.4220    | 7.2868    | 18.8597   | 69.6273   | 192.3792  | 463.7308 |

### groupby.sum
| rows\cols | 32        | 64        | 128       | 256       | 512       | 1024      | 2048      | 3584     |
|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|----------|
| 10**4     | 0.3525    | 0.7186    | 1.1743    | 2.3160    | 4.8310    | 10.8509   | 33.6107   | 100.7768 |

# Previous results from main
### fillna
| rows\cols | 32        | 64        | 128       | 256       | 512       | 1024      | 2048      | 3584     |
|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|----------|
| 10**4     | 0.0489    | 0.0879    | 0.1756    | 0.4972    | 1.3735    | 3.7084    | 9.9866    | 26.9021  |
| 10**5     | 0.0772    | 0.1471    | 0.2382    | 0.5195    | 1.4396    | 3.6447    | 14.2589   | 37.6544  |
| 10**6     | 0.0902    | 0.1910    | 0.5250    | 1.3777    | 5.2725    | 20.2653   | 84.3823   | 268.2694 |

### dropna
| rows\cols | 32        | 64        | 128       | 256       | 512       | 1024      | 2048      | 3584     |
|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|----------|
| 10**4     | 0.2343    | 0.4562    | 0.7633    | 1.4865    | 3.7967    | 0.2784    | 0.5159    | 0.9843   |
| 10**5     | 0.3956    | 0.5020    | 0.9178    | 1.6317    | 3.2689    | 0.3431    | 0.5513    | 1.0118   |
| 10**6     | 0.3732    | 0.5757    | 1.2170    | 2.0218    | 4.5112    | 1.0296    | 1.8023    | 3.5953   |
| 10**7     | 0.3134    | 1.0215    | 1.9741    | 5.0844    | 10.6587   | 21.2115   | 85.1395   | 238.7371 |

### groupby.sum
| rows\cols | 32        | 64        | 128       | 256       | 512       | 1024      | 2048      | 3584     |
|-----------|-----------|-----------|-----------|-----------|-----------|-----------|-----------|----------|
| 10**4     | 0.3703    | 0.6839    | 1.3411    | 2.8394    | 6.7643    | 18.2716   | 64.4190   | 204.0969 |

# Explanatory note
As you can see, fillna and groupby results seem expected and this PR shows good speedup, but dropna results seem strange. This is because PR #6472 contained improvements for dropna, but issue#5347 affected it and the new approach only applied to dataframes that have less than 32 column partitions (1024 columns when using the default MinPartitionSize). This PR fixes that problem, but it doesn't help because the old way looks better for data frames with more columns. 
This can be seen by the fact that the performance results increase for data frames with less than 1024 columns, but larger dataframes still have better performance.

**Comment by anmyachev** [comment]
> This PR fixes that problem, but it doesn't help because the old way looks better for data frames with more columns.

@Retribution98 do you plan to preserve that path?

File: modin/core/dataframe/algebra/groupby.py [hunk 8]
```diff
@@ -655,9 +655,11 @@ def aggregate_on_dict(grp_obj, *args, **kwargs):
...
 
             native_res_part = [] if native_agg_res is None else [native_agg_res]
-            result = pandas.concat(
-                [*native_res_part, *custom_results], axis=1, copy=False
-            )
+            parts = [*native_res_part, *custom_results]
+            if parts:
+                result = pandas.concat(parts, axis=1, copy=False)
+            else:
+                result = pandas.DataFrame(columns=result_columns)
```
Comment (anmyachev) [inline modin/core/dataframe/algebra/groupby.py:662]: Is it important to keep the column names in case of empty partitions? How did you come to this?

Comment (Retribution98) [reply modin/core/dataframe/algebra/groupby.py:662]: I do it to get the expectable result, because result columns are known already. Would it bring any problems or slowdowns?

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 9]
```diff
@@ -3107,9 +3108,7 @@ def dropna(self, **kwargs):
             lib.no_default,
             None,
         )
-        # FIXME: this is a naive workaround for this problem: https://github.com/modin-project/modin/issues/5394
-        # if there are too many partitions then all non-full-axis implementations start acting very badly.
-        # The here threshold is pretty random though it works fine on simple scenarios
+        # The map reduce approach works well for frames with few columnar partitions
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:3111]:
```suggestion
        # The map reduce approach works well for frames with few row or column partitions
```

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 10]
```diff
@@ -3265,9 +3282,31 @@ def broadcast_apply(
...
 
-        new_frame = self._partition_mgr_cls.broadcast_apply(
-            axis, func, left_parts, right_parts
-        )
+        # check the conditions for use of dynamic partitioning
+        use_dynamic_partitioning = False
+        if self.num_parts <= 1.5 * CpuCount.get():
+            use_dynamic_partitioning = True
+
+            # When the frame is large, dynamic partitioning
```
Comment (YarShev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:3290]:
```suggestion
            # When the frame is "large", dynamic partitioning
```

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 11]
```diff
@@ -3265,9 +3282,31 @@ def broadcast_apply(
...
+            # performs worse than the based approach
```
Comment (YarShev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:3291]:
```suggestion
            # performs worse than the basic approach
```

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 12]
```diff
@@ -3265,9 +3282,31 @@ def broadcast_apply(
...
+            frame_size = self.size
+            if frame_size and (frame_size >= 4 * 10**9 or len(self) >= 10**7):
```
Comment (YarShev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:3293]: You introduced a new method .size, which avoid triggering computation for row and columns, but then use len(self). My questions are:

1) Should we keep the condition for len(self)?
2) Should we add the condition for len(self.columns)?
3) Should we have the new method .size()

Comment (YarShev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:3293]: I see, thanks! Should we have a check for len(self.columns)?

Comment (Retribution98) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:3293]: Probably, but I haven't checked because it's a rare case

File: modin/core/dataframe/pandas/partitioning/partition_manager.py (hunk 5 above)
Comment (Retribution98) [reply modin/core/dataframe/pandas/partitioning/partition_manager.py:686]: This condition changed using a new environment variable `DynamicPartitioning`

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 9 above)
Comment (Retribution98) [reply modin/core/storage_formats/pandas/query_compiler.py:3111]: It is not correct. I got the follows results:
If Dataframe has few rows (less than 10^7) the old way show a better performance regardless of the number of columns.
If Dataframe has plenty of rows (more than 10^7) and few columns the map reduce approach is better.
If Dataframe has plenty of rows (more than 10^7) and plenty of columns the old  approach is better again.

But I didn't add any new conditions here, because this is beyond the scope of the current task.

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 12 above)
Comment (Retribution98) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: Outdated

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 11 above)
Comment (Retribution98) [reply modin/core/dataframe/pandas/dataframe/dataframe.py]: Outdated

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 13]
```diff
@@ -45,7 +45,8 @@
 from pandas.core.indexing import check_bool_indexer
 from pandas.errors import DataError
 
-from modin.config import CpuCount, RangePartitioning
+from modin.config import RangePartitioning
+from modin.config.envvars import CpuCount
```
[the remark this answers is not in the document]
Comment (Retribution98) [reply modin/core/storage_formats/pandas/query_compiler.py]: Done

File: modin/core/dataframe/algebra/groupby.py (hunk 8 above)
Comment (anmyachev) [reply modin/core/dataframe/algebra/groupby.py:662]: I just want to make sure that Modin's behavior matches that of a pandas. Do we have a test for this new code?

**Review by anmyachev — DISMISSED** [review]
LGTM!

File: modin/tests/pandas/test_groupby.py [hunk 14]
```diff
@@ -2431,6 +2437,32 @@ def test_multi_column_groupby_different_partitions(
     )
 
 
+def test_empty_partitions_after_groupby():
+    DynamicPartitioning.put(True)
```
Comment (anmyachev) [inline modin/tests/pandas/test_groupby.py:2441]: Let's use context manager to be safe

Comment (Retribution98) [reply modin/tests/pandas/test_groupby.py]: sure, thanks

**Review by anmyachev — APPROVED** [review]
